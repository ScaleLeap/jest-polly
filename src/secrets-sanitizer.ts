import merge from 'lodash.merge'
import { isDefined } from 'ts-is-present'
import { JsonObject } from 'type-fest'

import { Secrets } from './config'

type NormalizedSecrets = Record<string, string>

export function replaceAll(string: string, secrets: NormalizedSecrets): string {
  let secretString = string

  for (const [secret, replacer] of Object.entries(secrets)) {
    secretString = secretString.split(secret).join(replacer)
  }

  return secretString
}

export function normalizeSecrets(secrets: Secrets): NormalizedSecrets {
  if (!Array.isArray(secrets)) {
    return secrets
  }

  const normalizedSecrets: NormalizedSecrets = {}

  const definedSecrets = secrets
    .filter(isDefined)
    .filter((secret) => typeof secret === 'string' && !!secret)

  for (const secret of definedSecrets) {
    Object.assign(normalizedSecrets, { [secret]: 'x' })
  }

  return normalizedSecrets
}

// eslint-disable-next-line radar/cognitive-complexity
export function sanitize(recording: JsonObject, secrets: NormalizedSecrets): JsonObject {
  const accumulator: JsonObject = merge({}, recording)

  for (const key of Object.keys(recording)) {
    const value = accumulator[key]

    if (typeof value === 'string') {
      accumulator[key] = replaceAll(value, secrets)
    }

    if (typeof value === 'object' && !Array.isArray(value) && value !== null) {
      accumulator[key] = sanitize(value, secrets)
    }

    if (Array.isArray(value)) {
      for (const [index, value_] of value.entries()) {
        if (typeof value_ === 'string') {
          value[index] = replaceAll(value_, secrets)
        }

        if (typeof value_ === 'object' && !Array.isArray(value_) && value_ !== null) {
          value[index] = sanitize(value_, secrets)
        }
      }
    }
  }

  return accumulator
}

export function secretSanitizer(recording: JsonObject, secrets: Secrets): JsonObject {
  return sanitize(merge({}, recording), normalizeSecrets(secrets))
}
