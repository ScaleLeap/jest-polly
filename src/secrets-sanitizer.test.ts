import { normalizeSecrets, replaceAll, secretSanitizer } from './secrets-sanitizer'

describe('secrets', () => {
  describe(`${replaceAll.name}`, () => {
    it('should replace nothing if empty secrets are given', () => {
      expect.assertions(1)

      expect(replaceAll('foo', {})).toBe('foo')
    })

    it('should replace one secret', () => {
      expect.assertions(1)

      expect(replaceAll('foo iamseret bar', { iamseret: 'x' })).toBe('foo x bar')
    })

    it('should replace two secrets', () => {
      expect.assertions(1)

      const secrets = {
        iamseret: 'x',
        anothersecret: 'y',
      }

      expect(replaceAll('foo iamseret bar anothersecret baz', secrets)).toBe('foo x bar y baz')
    })
  })

  describe(`${normalizeSecrets.name}`, () => {
    it('should normalize array', () => {
      expect.assertions(1)

      const normalized = normalizeSecrets(['a', 'b'])

      expect(normalized).toStrictEqual({
        a: 'x',
        b: 'x',
      })
    })

    it('should return record-style secret param as is', () => {
      expect.assertions(1)

      const secrets = {
        a: 'x',
        b: 'x',
      }

      expect(normalizeSecrets(secrets)).toStrictEqual(secrets)
    })
  })

  describe(`${secretSanitizer.name}`, () => {
    const secrets = {
      iamasecret: 'x',
      anothersecret: 'x',
    }

    it('should replace all secrets and leave other params', () => {
      expect.assertions(1)

      const recording = {
        foo: '1 iamasecret 1',
        bar: {
          // eslint-disable-next-line unicorn/no-null
          null: null,
          baz: {
            foo: true,
            val: '1 anothersecret 1 anothersecret x',
          },
        },
        array: [
          {
            name: 'foo',
            value: 'iamasecret',
          },
        ],
        arrayOfStrings: ['iamasecret', 'y'],
      }

      expect(secretSanitizer(recording, secrets)).toMatchSnapshot()
    })
  })
})
