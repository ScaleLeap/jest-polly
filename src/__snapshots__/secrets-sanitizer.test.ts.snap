// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`secrets secretSanitizer should replace all secrets and leave other params 1`] = `
{
  "array": [
    {
      "name": "foo",
      "value": "x",
    },
  ],
  "arrayOfStrings": [
    "x",
    "y",
  ],
  "bar": {
    "baz": {
      "foo": true,
      "val": "1 x 1 x x",
    },
    "null": null,
  },
  "foo": "1 x 1",
}
`;
